k256/src/arithmetic/field/field_impl.rs

```diff
@@ -0,0 +1,209 @@
+        CtOption::map(value, |x| Self::new_normalized(&x))
+    }
+
+    /// Returns the SEC-1 encoding of this field element.
+    pub fn to_bytes(&self) -> [u8; 32] {
+        debug_assert!(self.normalized);
+        self.value.to_bytes()
+    }
+
+    pub fn normalize_weak(&self) -> Self {
+        Self::new_weak_normalized(&self.value.normalize_weak())
+    }
+
+    pub fn normalize(&self) -> Self {
+        Self::new_normalized(&self.value.normalize())
+    }
+
+    pub fn normalizes_to_zero(&self) -> Choice {
+        self.value.normalizes_to_zero()
+    }
+
+    pub fn to_words(&self) -> [u64; 4] {
+        self.normalize().value.to_words()
+    }
+
+    pub const fn from_words_unchecked(words: [u64; 4]) -> Self {
+        let value = FieldElementUnsafeImpl::from_words_unchecked(words);
+        Self::new_normalized(&value)
+    }
+
+    pub fn from_words(words: [u64; 4]) -> CtOption<Self> {
+        let value = FieldElementUnsafeImpl::from_words(words);
+        CtOption::map(value, |x| Self::new_normalized(&x))
+    }
+
+    pub fn is_zero(&self) -> Choice {
+        debug_assert!(self.normalized);
+        self.value.is_zero()
+    }
+
+    pub fn is_odd(&self) -> Choice {
+        debug_assert!(self.normalized);
+        self.value.is_odd()
+    }
+
+    pub fn negate(&self, magnitude: u32) -> Self {
+        debug_assert!(self.magnitude <= magnitude);
+        let new_magnitude = magnitude + 1;
+        debug_assert!(new_magnitude <= FieldElementUnsafeImpl::max_magnitude());
+        Self::new(&(self.value.negate(magnitude)), new_magnitude)
+    }
+
+    pub fn add(&self, rhs: &Self) -> Self {
+        let new_magnitude = self.magnitude + rhs.magnitude;
+        debug_assert!(new_magnitude <= FieldElementUnsafeImpl::max_magnitude());
+        Self::new(&(self.value.add(&(rhs.value))), new_magnitude)
+    }
+
+    pub fn double(&self) -> Self {
+        let new_magnitude = self.magnitude * 2;
+        debug_assert!(new_magnitude <= FieldElementUnsafeImpl::max_magnitude());
+        Self::new(&(self.value.double()), new_magnitude)
+    }
+
+    pub fn mul_single(&self, rhs: u32) -> Self {
+        let new_magnitude = self.magnitude * rhs;
+        debug_assert!(new_magnitude <= FieldElementUnsafeImpl::max_magnitude());
+        Self::new(&(self.value.mul_single(rhs)), new_magnitude)
+    }
+
+    /// Returns self * rhs mod p
+    pub fn mul(&self, rhs: &Self) -> Self {
+        debug_assert!(self.magnitude <= 8);
+        debug_assert!(rhs.magnitude <= 8);
+        Self::new_weak_normalized(&(self.value.mul(&(rhs.value))))
+    }
+
+    /// Returns self * self mod p
+    pub fn square(&self) -> Self {
+        debug_assert!(self.magnitude <= 8);
+        Self::new_weak_normalized(&(self.value.square()))
+    }
+}
+
+#[cfg(debug_assertions)]
+impl Default for FieldElementImpl {
+    fn default() -> Self {
+        Self::zero()
+    }
+}
+
+#[cfg(debug_assertions)]
+#[cfg(test)]
```

Comment:
`debug_assertions` defaults to on for the `test` profile, so this feels a little redundant:

https://doc.rust-lang.org/cargo/reference/profiles.html#test

It seems like you could either just use `#[cfg(test)]` or something like the following:

```suggestion
#[cfg(all(debug_assertions, test))]
```

Reply:
Somewhat fixed, I guess. Much less `#[cfg(test)]` around the code now.

Reply:
Also, I think it is worth it to keep `debug_assertions` and `test` separate. `test` marks whatever is used in tests, `debug_assertions` marks debug assertions. These are orthogonal flags; just because by default `debug_assertions` is on when `tests` is on, does not mean that they can't be active separately under some conditions.